pod2daemon/csidriver/driver/node.go

```diff
@@ -0,0 +1,313 @@
+		}
+		e = os.RemoveAll(newDir)
+		if e != nil {
+			log.Errorf("Mount error: failed to clear %s: %v", newDir, e)
+		}
+		log.Errorf("Failed to create nodeagent directory at %s: %v", newDestinationDir, err)
+		return err
+	}
+
+	// Do a bind mount
+	// Run "mount --bind newDir newDestinationDir"
+	err = syscall.Mount(newDir, newDestinationDir, "", syscall.MS_BIND, "")
+	if err != nil {
+		// Run "umount destinationDir"
+		e := syscall.Unmount(destinationDir, 0)
+		if e != nil {
+			log.Errorf("Mount error: failed to unmount %s: %v", destinationDir, e)
+		}
+		e = os.RemoveAll(newDir)
+		if e != nil {
+			log.Errorf("Mount error: failed to clear %s: %v", newDir, e)
+		}
+		log.Errorf("Failed to bind mount %s to %s: %v", newDir, newDestinationDir, err)
+		return err
+	}
+
+	return nil
+}
+
+func (ns *nodeService) addCredentialFile(volumeID string, podInfo *creds.Credentials) error {
+	//Make the directory and then write the podInfo as json to it.
+	err := os.MkdirAll(ns.config.NodeAgentCredentialsHomeDir, 0755)
+	if err != nil {
+		return err
+	}
+
+	var attrs []byte
+	attrs, err = json.Marshal(podInfo)
+	if err != nil {
+		return err
+	}
+
+	credsFileTmp := strings.Join([]string{ns.config.NodeAgentManagementHomeDir, volumeID + ".json"}, "/")
+	_ = ioutil.WriteFile(credsFileTmp, attrs, 0644)
+
+	// Move it to the right location now.
+	credsFile := strings.Join([]string{ns.config.NodeAgentCredentialsHomeDir, volumeID + ".json"}, "/")
+	return os.Rename(credsFileTmp, credsFile)
+}
+
+func (ns *nodeService) removeCredentialFile(volumeId string) error {
+	credsFile := strings.Join([]string{ns.config.NodeAgentCredentialsHomeDir, volumeId + ".json"}, "/")
+	err := os.Remove(credsFile)
+	return err
+}
+
+func (ns *nodeService) retrievePodInfoFromFile(volumeId string) (*creds.Credentials, error) {
+	credsFilePath := strings.Join([]string{ns.config.NodeAgentCredentialsHomeDir, volumeId + ".json"}, "/")
+
+	credsFile, err := os.Open(credsFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	defer credsFile.Close()
+
+	credsFileBytes, err := ioutil.ReadAll(credsFile)
+	if err != nil {
+		return nil, err
+	}
+
+	podInfo := creds.Credentials{}
+	err = json.Unmarshal(credsFileBytes, &podInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	return &podInfo, nil
+}
+
+func (ns *nodeService) unmount(dir, volumeId string) error {
+	// unmount the bind mount
+	err := unmount(dir + "/nodeagent")
+	if err != nil {
+		log.Errorf("Unmount error: failed to unmount %s/nodeagent: %v", dir, err)
+	}
+
+	// unmount the tmpfs
+	err = unmount(dir)
+	if err != nil {
+		log.Errorf("Unmount error: failed to unmount %s: %v", dir, err)
+	}
+
+	// delete the directory that was created.
+	delDir := strings.Join([]string{ns.config.NodeAgentWorkloadHomeDir, volumeId}, "/")
+	err = os.RemoveAll(delDir)
+	if err != nil {
+		log.Errorf("Unmount error: unable to remove mount directory %s: %v", delDir, err)
+		return err
+	}
+
+	return nil
+}
+
+func unmount(dir string) error {
```

Comment:
Why do we need a wrapper function for this? Seems to basically just call one function and return its error.